controller/src/main/java/io/pravega/controller/store/PravegaTablesStoreHelper.java

```diff
@@ -244,6 +249,89 @@ public void invalidateCache(String table, String key) {
                 .whenComplete((r, ex) -> releaseEntries(entries));
     }
 
+    public CompletableFuture<List<Version>> removeValues(final String tableName, Set<String> tableKeys, final String appendValue, long requestId) {
+        Supplier<String> errorMessage = () -> String.format("remove values: on table: %s", tableName);
+
+        // read values and add values
+        return withRetries(() -> removeValuesTables(tableName, appendValue, tableKeys, requestId)
+                .thenApplyAsync(x -> {
+                    log.trace("appendValues {} updated to table {} with version {}", appendValue, tableName, x);
+                    return x.stream().map(v -> (Version) new Version.LongVersion(v.getSegmentVersion())).collect(Collectors.toList());
+                }), errorMessage, true, requestId);
+    }
+
+
+    public CompletableFuture<List<Version>> appendValues(final String tableName, Set<String> tableKeys, final String appendValue, long requestId) {
+        Supplier<String> errorMessage = () -> String.format("append values: on table: %s", tableName);
+
+        // read values and add values
+        return withRetries(() -> appendValuesTable(tableName, appendValue, tableKeys, requestId)
+                .thenApplyAsync(x -> {
+                    log.trace("appendValues {} updated to table {} with version {}", appendValue, tableName, x);
+                    return x.stream().map(v -> (Version) new Version.LongVersion(v.getSegmentVersion())).collect(Collectors.toList());
+                }), errorMessage, true, requestId);
+    }
+
+    private CompletableFuture<List<TableSegmentKeyVersion>> removeValuesTables(String tableName, String removeValue, Set<String> tableKeys, long requestId) {
+        // Get unversioned keys.
+        final List<TableSegmentKey> keys = tableKeys.stream()
+                                                    .map(k -> TableSegmentKey.unversioned(k.getBytes(StandardCharsets.UTF_8)))
+                                                    .collect(Collectors.toList());
+
+        CompletableFuture<List<TableSegmentEntry>> f1 = segmentHelper.readTable(tableName, keys, authToken.get(), requestId) // read the latest version of keys.
+                                                                     .whenComplete((v, ex) -> releaseKeys(keys));
+        f1.thenCompose(currentEntries -> {
+            List<TableSegmentEntry> updatedList = currentEntries.stream().map(entry -> {
+                String k = entry.getKey().getKey().toString(Charset.defaultCharset());
+                byte[] array = getArray(entry.getValue());
+                byte[] updatedBytes;
+                if (array.length == 0) {
+                    updatedBytes = TagRecord.builder().tagName(k).build().toBytes();
```

Comment:
again, it is assuming that the value is of type tagRecord. this is very business logic specific and should go into tag specific caller class.

Reply:
done.